examples/games/desktoy.rs

```diff
@@ -0,0 +1,320 @@
+                                    mesh: circle.clone(),
+                                    material: pupil_highlight_material.clone(),
+                                    transform: Transform::from_xyz(
+                                        -pupil_highlight_offset,
+                                        pupil_highlight_offset,
+                                        1.0,
+                                    )
+                                    .with_scale(Vec3::new(
+                                        pupil_highlight_radius,
+                                        pupil_highlight_radius,
+                                        1.0,
+                                    )),
+                                    ..default()
+                                });
+                            });
+                    });
+            }
+        });
+}
+
+/// Project the cursor into the world coordinates and store it in a resource for easy use
+fn get_cursor_world_pos(
+    mut cursor_world_pos: ResMut<CursorWorldPos>,
+    q_primary_window: Query<&Window, With<PrimaryWindow>>,
+    q_main_camera: Query<(&Camera, &GlobalTransform), With<MainCamera>>,
+) {
+    let primary_window = q_primary_window.single();
+    let (main_camera, main_camera_transform) = q_main_camera.single();
+    // Get the cursor position in the world
+    cursor_world_pos.0 = primary_window
+        .cursor_position()
+        .and_then(|cursor_pos| main_camera.viewport_to_world_2d(main_camera_transform, cursor_pos));
+}
+
+/// Update whether the window is clickable or not
+fn update_cursor_hit_test(
+    cursor_world_pos: Res<CursorWorldPos>,
+    mut q_primary_window: Query<&mut Window, With<PrimaryWindow>>,
+    q_bevy_logo: Query<&Transform, With<BevyLogo>>,
+) {
+    let mut primary_window = q_primary_window.single_mut();
+    if primary_window.decorations {
+        primary_window.cursor.hit_test = true;
+        return;
+    }
+
+    let Some(cursor_world_pos) = cursor_world_pos.0 else {
+        return;
+    };
+
+    let bevy_logo_transform = q_bevy_logo.single();
+    primary_window.cursor.hit_test =
+        (bevy_logo_transform.translation.truncate() - cursor_world_pos).length() < BEVY_LOGO_RADIUS;
+}
+
+/// Start the drag operation and record the offset we started dragging from
+fn start_drag(
+    mut commands: Commands,
+    cursor_world_pos: Res<CursorWorldPos>,
+    q_bevy_logo: Query<&Transform, With<BevyLogo>>,
+) {
+    let Some(cursor_world_pos) = cursor_world_pos.0 else {
+        return;
+    };
+
+    let bevy_logo_transform = q_bevy_logo.single();
+    let drag_offset = bevy_logo_transform.translation.truncate() - cursor_world_pos;
+    if drag_offset.length() < BEVY_LOGO_RADIUS {
+        commands.insert_resource(DragOffset(drag_offset));
+    }
+}
+
+/// Stop the current drag operation
+fn end_drag(mut commands: Commands) {
+    commands.remove_resource::<DragOffset>();
```

Comment:
Can we model this using a state machine on an enum instead? I tend to find that pattern much clearer to understand.

Reply:
I didn't turn the drag operation into a state machine since there is no cleaner way to test if there is a current drag operation while also storing the drag offset. E.g. there is no resource_equals equivalent that would ignore the inner enum values.
However I renamed DragOffset to DragOperation to make it clearer that this resource represents the drag operation.